File: Loenn/entities/battery_gate.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local battery_gate = {}

battery_gate.name = "batteries/battery_gate"
battery_gate.depth = -9000

battery_gate.placements = {
    {
        name = "vertical",
        data = {
            switchId = -1,
            vertical = true,
            closes = false
        }
    },
    {
        name = "vertical_closing",
        data = {
            switchId = -1,
            vertical = true,
            closes = true
        }
    },
    {
        name = "horizontal",
        data = {
            switchId = -1,
            vertical = false,
            closes = false
        }
    },
    {
        name = "horizontal_closing",
        data = {
            switchId = -1,
            vertical = false,
            closes = true
        }
    },
}

battery_gate.fieldInformation = {
    switchId = {
        fieldType = "integer"
    }
}

battery_gate.canResize = {false, false} -- prevent resizing gates placed with old Ahorn plugin

function battery_gate.sprite(room, entity)
    local texture = (entity.closes ? "batteries/battery_gate/door15" : "batteries/battery_gate/door1")
    local sprite = drawableSprite.fromTexture(texture, entity)

    if entity.vertical then
        sprite:addPosition(7, 24)
    else
        sprite:addPosition(24, 8)
        sprite.rotation = -math.pi / 2
    end

    return sprite
end

function battery_gate.selection(room, entity)
    if entity.vertical then
        return utils.rectangle(entity.x, entity.y, 15, 48)
    else
        return utils.rectangle(entity.x, entity.y, 48, 15)
    end
end

return battery_gate

File: Ahorn/entities/battery_gate.jl
module BatteriesBatteryGate

using ..Ahorn, Maple

@mapdef Entity "batteries/battery_gate" BatteryGate(x::Integer, y::Integer, vertical::Bool=true, closes::Bool=false, switchId::Integer=-1)

const placements = Ahorn.PlacementDict(
    "Battery Gate (Vertical) (Batteries)" => Ahorn.EntityPlacement(
        BatteryGate,
        "point",
        Dict{String, Any}(
            "vertical" => true,
        )
    ),
    "Battery Gate (Horizontal) (Batteries)" => Ahorn.EntityPlacement(
        BatteryGate,
        "point",
        Dict{String, Any}(
            "vertical" => false,
        )
    ),
    "Closing Battery Gate (Vertical) (Batteries)" => Ahorn.EntityPlacement(
        BatteryGate,
        "point",
        Dict{String, Any}(
            "vertical" => true,
            "closes" => true,
        )
    ),
    "Closing Battery Gate (Horizontal) (Batteries)" => Ahorn.EntityPlacement(
        BatteryGate,
        "point",
        Dict{String, Any}(
            "vertical" => false,
            "closes" => true,
        )
    ),
)

function Ahorn.selection(entity::BatteryGate)
    x, y = Ahorn.position(entity)
    height = Int(get(entity.data, "height", 48))

    if get(entity.data, "vertical", true)
        return Ahorn.Rectangle(x, y, 15, height)
    else
        return Ahorn.Rectangle(x, y, height, 15)
    end
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::BatteryGate, room::Maple.Room)
    if get(entity.data, "closes", false)
        sprite = "batteries/battery_gate/door15"
    else
        sprite = "batteries/battery_gate/door1"
    end

    if get(entity.data, "vertical", true)
        Ahorn.drawSprite(ctx, sprite, 7, 24)
    else
        Ahorn.drawSprite(ctx, sprite, 12, 44, rot=-pi / 2)
    end
end

end
